### PR TITLE
Install by version and upgrade all on new releases

### PR DESCRIPTION
**Pull request by lewisdiamond**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1alpha1/packagebundle.go [hunk 1]
```diff
@@ -23,20 +25,22 @@ func (config *PackageBundle) ExpectedKind() string {
...
 	for _, pkg := range config.Spec.Packages {
 		if pkg.Name == pkgName {
 			source := pkg.Source
 			for _, version := range source.Versions {
-				if version.Name == pkgVersion || version.Digest == pkgVersion {
-					retSource = PackageOCISource{Registry: source.Registry, Repository: source.Repository, Digest: version.Digest}
-					return retSource, version.Name, nil
+				//We do not sort before getting `latest` because there will be only a single version per release in normal cases. For edge cases which may require multiple
+				//versions, the order in the file will be ordered according to what we want `latest` to point to
+				if version.Name == pkgVersion || version.Digest == pkgVersion || pkgVersion == Latest {
```
Comment (ewollesen) [inline api/v1alpha1/packagebundle.go:35]: Not asking you to change this, but should we be doing a case-insensitive comparison? At least for the digest, case definitely doesn't matter, and I think we can make a case that case shouldn't matter in either of the other cases as well.

Are there any cases where case would matter for this purpose? I can't think of any. Also, at least in the case of a digest, I can imagine someone specifying the digest in whatever case we're not expecting.

File: api/v1alpha1/packagebundle_test.go [hunk 2]
```diff
@@ -11,45 +11,96 @@ import (
...
 
-	actual, version, err = sut.FindSource("eks-anywhere-test", "sha256:eaa07ae1c06ffb563fe3c16cdb317f7ac31c8f829d5f1f32442f0e5ab982c3e7")
+	actual, err = sut.FindSource("eks-anywhere-test", "sha256:eaa07ae1c06ffb563fe3c16cdb317f7ac31c8f829d5f1f32442f0e5ab982c3e7")
 	assert.NoError(t, err)
 	assert.Equal(t, expected, actual)
-	assert.Equal(t, expectedVersion, version)
 
-	expectedErr := "package not found: Bogus @ bar"
-	_, _, err = sut.FindSource("Bogus", "bar")
+	expectedErr := "package not found in current active bundle: Bogus @ bar"
```
Comment (ewollesen) [inline api/v1alpha1/packagebundle_test.go:53]:
```suggestion
	expectedErr := "package not found in active bundle (): Bogus @ bar"
```

File: controllers/package_controller.go [hunk 3]
```diff
@@ -124,8 +144,19 @@ func (r *PackageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+		targetVersion := managerContext.Package.Spec.PackageVersion
+		if targetVersion == "" {
+			targetVersion = api.Latest
+		}
+		pkgName := managerContext.Package.Spec.PackageName
+		managerContext.Source, err = bundle.FindSource(pkgName, targetVersion)
+		managerContext.Package.Status.TargetVersion = printableTargetVersion(managerContext.Source, targetVersion)
 		if err != nil {
+			//TODO add a link to documentation on how to make a bundle active.
+			managerContext.Package.Status.Detail = fmt.Sprintf("Package %s@%s is not in the current active bundle. Did you forget to activate the new bundle?", pkgName, targetVersion)
```
Comment (ewollesen) [inline controllers/package_controller.go:156]:
```suggestion
			managerContext.Package.Status.Detail = fmt.Sprintf("Package %s@%s is not in the active bundle (%s). Did you forget to activate the new bundle?", bundle.Version, pkgName, targetVersion)
```

I wouldn't output the second sentence unless I knew there was a newer bundle on the system. I feel like error messages that are misleading are worse than terse ones.

File: controllers/package_controller_test.go [hunk 4]
```diff
@@ -165,6 +170,89 @@ func TestReconcile(t *testing.T) {
...
+			DoAndReturn(fn)
+
+		tf.bundleManager.FakeActiveBundle = tf.mockBundle()
+
+		testErr := errors.New("status update test error")
+		status := tf.mockStatusWriter()
+
+		pkg.Spec.PackageVersion = "2.0.0"
+		pkg.Status.TargetVersion = "2.0.0"
+		pkg.Status.Detail = fmt.Sprintf("Package %s@%s is not in the current active bundle. Did you forget to activate the new bundle?", pkg.Spec.PackageName, pkg.Spec.PackageVersion)
```
Comment (ewollesen) [inline controllers/package_controller_test.go:189]:
```suggestion
		pkg.Status.Detail = fmt.Sprintf("Package %s@%s is not in the active bundle (). Did you forget to activate the new bundle?", pkg.Spec.PackageName, pkg.Spec.PackageVersion)
```
Heh, there's no bundle version in `tf.mockBundle()`

File: api/v1alpha1/packagebundle.go [hunk 5]
```diff
@@ -23,20 +25,22 @@ func (config *PackageBundle) ExpectedKind() string {
...
+					retSource = PackageOCISource{Registry: source.Registry, Repository: source.Repository, Digest: version.Digest, Version: version.Name}
+					return retSource, nil
 				}
 			}
 		}
 	}
 
-	return retSource, "", fmt.Errorf("package not found: %s @ %s", pkgName, pkgVersion)
+	return retSource, fmt.Errorf("package not found in current active bundle: %s @ %s", pkgName, pkgVersion)
```
Comment (ewollesen) [inline api/v1alpha1/packagebundle.go:43]:
```suggestion
	return retSource, fmt.Errorf("package not found in active bundle (%s): %s @ %s", config.Spec.Version, pkgName, pkgVersion)
```

I've made a couple of suggestions in line with this one, but just via the web interface, so I may well have goofed something. Hopefully they're more useful than not though.

File: api/v1alpha1/packagebundle_test.go [hunk 6]
```diff
@@ -11,45 +11,96 @@ import (
...
+	givenBundle := func(versions []api.SourceVersion) api.PackageBundle {
+		return api.PackageBundle{
+			Spec: api.PackageBundleSpec{
+				Packages: []api.BundlePackage{
+					{
+						Name: "eks-anywhere-test",
+						Source: api.BundlePackageSource{
+							Registry:   "public.ecr.aws/l0g8r8j6",
+							Repository: "eks-anywhere-test",
+							Versions:   versions},
```
Comment (ewollesen) [inline api/v1alpha1/packagebundle_test.go:23]:
```suggestion
							Versions:   versions,
					        },
```
Very much a nit, but....

**Review by ewollesen — APPROVED** [review]
Just small ideas/tweaks from me. I can be persuaded to merge as-is though.

File: api/v1alpha1/packagebundle_test.go (hunk 6 above)
Comment (lewisdiamond) [reply api/v1alpha1/packagebundle_test.go:23]: Isn't that supposed to get linted? I did run it with `--fix`

Comment (ewollesen) [reply api/v1alpha1/packagebundle_test.go]: I don't think so, strangely enough... There are occasions where you might leave it this way, e.g.:
```go
foo := myType{ foo: "bar",
    ack: "gar"}
```
Notice the assignments start on the first line. I don't love this style, but sometimes it makes sense.

**Comment by lewisdiamond** [comment]
> @lewisdiamond: you cannot LGTM your own PR.
Well seems like you're alive.

**Pull request merged**